pkg/lib/credentials.js

```diff
@@ -180,78 +180,78 @@ function Keys() {
         return cockpit.script('dir=$(dirname "$1"); test -e "$dir" || mkdir -m 700 "$dir"', [file]);
     }
 
-    function run_keygen(file, new_type, old_pass, new_pass, two_pass) {
+    async function run_keygen(file, new_type, old_pass, new_pass, two_pass) {
         const old_exps = [/.*Enter old passphrase: $/];
         const new_exps = [/.*Enter passphrase.*/, /.*Enter new passphrase.*/, /.*Enter same passphrase again: $/];
         const bad_exps = [/.*failed: passphrase is too short.*/];
 
-        return new Promise((resolve, reject) => {
-            let buffer = "";
-            let sent_new = false;
-            let failure = _("No such file or directory");
+        let buffer = "";
+        let sent_new = false;
+        let failure = _("No such file or directory");
+
+        if (new_pass !== two_pass)
+            return Promise.reject(new Error(_("The passwords do not match.")));
+
+        // Exactly one of new_type or old_pass must be given
+        console.assert((new_type == null) != (old_pass == null));
+
+        const cmd = ["ssh-keygen", "-f", file];
+        if (new_type)
+            cmd.push("-t", new_type);
+        else
+            cmd.push("-p");
+
+        await self.p_have_path;
+
+        const proc = cockpit.spawn(cmd, { pty: true, environ: ["LC_ALL=C"], err: "out", directory: self.path });
```

Comment:
The data flow between the `await` above and the use of `self.path` here is weird.  If you consider that `await` is just a wrapper around `.then()` then you're basically relying on the fact that the other `.then()` runs before our `.then()`, which ... sure... I think that's guaranteed, but why not make it more explicit?

Something like

`user = await self.p_have_path`

and then

`directory: user.home + '/.ssh'`

On the other hand, I guess your intention is that the `p_have_path` is a promise that explicitly resolves when the other `.then()` handler has run...

Reply:
With that we'd have to copy the definition of `directory` multiple times (once per method), which is also awkward. It could be another helper method, but that's exactly what promise chaining does -- once `p_have_path` resolves, `self.user` is guaranteed to be set.

Reply:
OK.